base/server/src/main/java/com/netscape/cmscore/ldapconn/LdapAnonConnFactory.java

```diff
@@ -410,25 +409,38 @@ public synchronized void returnConn(LDAPConnection conn) {
         for (int i = 0; i < mNumConns; i++) {
             if (mConns[i] == anon) {
                 logger.warn("LdapAnonConnFactory: Connection already returned");
+                --mTotal;
+                notifyAll();
+                return;
             }
         }
 
-        // this returned connection might authenticate as someone other than
-        // anonymonus. Reset it to anonymous first before it returns
-        // to the pool.  Do this by calling connect() again on this connection
-        // to avoid doing an explicit anonymous bind
-        try {
-            anon.connect(mConnInfo.getHost(), mConnInfo.getPort());
-        } catch (LDAPException e) {
-            logger.warn("LdapAnonConnFactory: Unable to reauthenticate as anonymous");
-        }
+        if(mNumConns < mMinConns) {
+            // this returned connection might authenticate as someone other than
+            // anonymonus. Reset it to anonymous first before it returns
+            // to the pool.  Do this by calling connect() again on this connection
+            // to avoid doing an explicit anonymous bind
+            try {
+                anon.connect(mConnInfo.getHost(), mConnInfo.getPort());
+            } catch (LDAPException e) {
+                logger.warn("LdapAnonConnFactory: Unable to reauthenticate as anonymous");
+            }
 
-        // return the connection even if can't reauthentication anon.
-        // most likely server was down.
-        mConns[mNumConns++] = anon;
+            // return the connection even if can't reauthentication anon.
+            // most likely server was down.
+            mConns[mNumConns++] = anon;
+        } else {
+            try {
+                anon.disconnect();
```

Comment:
I guess it's ok to disconnect since there's already a minimum number of connections in the pool. Perhaps in the future we can add some kind of delay so the connection can be reused quickly in case there's a lot of traffic.

Reply:
This to avoid connection stay on the pool for long time. If you have a burst increasing the pool size, then there was no way to reduce the size when returning to the normal activity. We can improve on this.